File: BEYLE MOBILE/plugins/jaal-browser/android/src/main/java/com/beyle/jaal/JaalBrowserPlugin.java
package com.beyle.jaal;

// BEYLE MOBILE/plugins/jaal-browser — Slice S19
//
// Capacitor plugin that manages a native Android WebView overlaid on top of
// the Capacitor WebView. Provides full browser control for JAAL research
// browsing: navigation, content extraction, JavaScript injection, and
// privacy enforcement (cookie blocking, request interception, storage isolation).
//
// The WebView is lazily created on the first navigate() call and added to the
// activity's root FrameLayout above Capacitor's WebView. It starts hidden
// (View.GONE) and is shown/hidden via show()/hide() or automatically on navigate().
//
// All @PluginMethod calls that touch the WebView are dispatched to the UI thread
// via getBridge().executeOnMainThread() — Android WebView MUST be manipulated
// on the UI thread only.

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.CookieManager;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.FrameLayout;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONArray;
import org.json.JSONException;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * JAAL Browser Capacitor Plugin — Slice S19.
 *
 * Exposes a native Android WebView with full privacy control to the
 * JAAL React UI components via the Capacitor JS bridge.
 */
@CapacitorPlugin(name = "JaalBrowser")
public class JaalBrowserPlugin extends Plugin {

    // ---- State ----
    private WebView jaalWebView;
    private JaalWebViewClient jaalClient;
    private PrivacyConfigHolder privacyConfig;
    private boolean isVisible = false;

    // ---- Privacy config holder ----

    /**
     * Holds parsed privacy configuration values.
     * Updated via setPrivacyConfig() from JS.
     * Read by JaalWebViewClient for request interception decisions.
     */
    static class PrivacyConfigHolder {
        boolean blockThirdPartyCookies = true;
        List<String> blockedDomains = new ArrayList<>();
        List<String> readOnlyDomains = new ArrayList<>();
        boolean storageIsolation = false;
        boolean disableJavaScript = false;
        String userAgent = null;
        boolean blockGeolocation = true;

        /**
         * Check if a URL's host matches any entry in blockedDomains.
         * Matching is by exact host or suffix (e.g., "ad.example.com" matches "example.com").
         */
        boolean isDomainBlocked(String url) {
            if (blockedDomains.isEmpty()) return false;
            String host = extractHost(url);
            if (host == null) return false;
            for (String domain : blockedDomains) {
                if (host.equals(domain) || host.endsWith("." + domain)) {
                    return true;
                }
            }
            return false;
        }

        /**
         * Return the matched blocked domain rule for a URL, or null if not blocked.
         */
        String getBlockedDomainMatch(String url) {
            String host = extractHost(url);
            if (host == null) return null;
            for (String domain : blockedDomains) {
                if (host.equals(domain) || host.endsWith("." + domain)) {
                    return domain;
                }
            }
            return null;
        }

        /**
         * Check if a URL's host matches any entry in readOnlyDomains.
         */
        boolean isDomainReadOnly(String url) {
            if (readOnlyDomains.isEmpty()) return false;
            String host = extractHost(url);
            if (host == null) return false;
            for (String domain : readOnlyDomains) {
                if (host.equals(domain) || host.endsWith("." + domain)) {
                    return true;
                }
            }
            return false;
        }

        /**
         * Extract the lowercase host from a URL string.
         */
        private static String extractHost(String url) {
            try {
                URI uri = URI.create(url);
                String host = uri.getHost();
                return host != null ? host.toLowerCase(Locale.ROOT) : null;
            } catch (Exception e) {
                return null;
            }
        }
    }

    // ---- Plugin lifecycle ----

    @Override
    public void load() {
        privacyConfig = new PrivacyConfigHolder();
    }

    @Override
    public void handleOnPause() {
        if (jaalWebView != null) {
            jaalWebView.onPause();
        }
    }

    @Override
    public void handleOnResume() {
        if (jaalWebView != null) {
            jaalWebView.onResume();
        }
    }

    @Override
    public void handleOnDestroy() {
        destroyWebView();
    }

    // ---- Lazy WebView creation ----

    /**
     * Create and configure the native WebView overlay if it doesn't already exist.
     * MUST be called on the UI thread.
     *
     * Layout hierarchy:
     *   android.R.id.content (FrameLayout)
     *     └─ Capacitor's WebView (index 0)
     *     └─ JAAL WebView (index 1, overlaid on top, starts GONE)
     */
    @SuppressLint("SetJavaScriptEnabled")
    private void ensureWebView() {
        if (jaalWebView != null) return;

        jaalWebView = new WebView(getContext());

        // --- WebSettings ---
        WebSettings settings = jaalWebView.getSettings();
        settings.setJavaScriptEnabled(!privacyConfig.disableJavaScript);
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);
        settings.setBuiltInZoomControls(true);
        settings.setDisplayZoomControls(false);
        settings.setLoadWithOverviewMode(true);
        settings.setUseWideViewPort(true);
        settings.setAllowContentAccess(false);
        settings.setAllowFileAccess(false);
        settings.setGeolocationEnabled(!privacyConfig.blockGeolocation);
        settings.setMediaPlaybackRequiresUserGesture(true);

        // Custom User-Agent for fingerprint defense
        if (privacyConfig.userAgent != null && !privacyConfig.userAgent.isEmpty()) {
            settings.setUserAgentString(privacyConfig.userAgent);
        }

        // --- Cookie management ---
        CookieManager cookieManager = CookieManager.getInstance();
        cookieManager.setAcceptCookie(true);
        cookieManager.setAcceptThirdPartyCookies(
            jaalWebView, !privacyConfig.blockThirdPartyCookies
        );

        // --- WebViewClient (request interception + navigation control + events) ---
        jaalClient = new JaalWebViewClient(this, privacyConfig);
        jaalWebView.setWebViewClient(jaalClient);

        // --- WebChromeClient (title tracking for pageLoad events) ---
        jaalWebView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onReceivedTitle(WebView view, String title) {
                // Title is tracked by the pageLoad event in JaalWebViewClient.onPageFinished
            }
        });

        // --- Layout: fill the overlay area ---
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        );
        jaalWebView.setLayoutParams(params);
        jaalWebView.setVisibility(View.GONE);
        jaalWebView.setBackgroundColor(Color.WHITE);

        // --- Add to the activity's root FrameLayout (above Capacitor WebView) ---
        ViewGroup rootView = (ViewGroup) getActivity()
            .findViewById(android.R.id.content);
        rootView.addView(jaalWebView);
    }

    // ---- @PluginMethod implementations ----

    @PluginMethod
    public void navigate(final PluginCall call) {
        final String url = call.getString("url");
        if (url == null || url.isEmpty()) {
            call.reject("url is required");
            return;
        }

        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                ensureWebView();
                jaalWebView.loadUrl(url);
                showOverlay(true);
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void getPageContent(final PluginCall call) {
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView == null) {
                    JSObject result = new JSObject();
                    result.put("html", "");
                    result.put("text", "");
                    result.put("url", "");
                    result.put("title", "");
                    call.resolve(result);
                    return;
                }

                // Extract HTML, text, URL, and title via JavaScript evaluation
                jaalWebView.evaluateJavascript(
                    "(function() { return JSON.stringify({ " +
                    "  html: document.documentElement.outerHTML, " +
                    "  text: document.body ? document.body.innerText : '', " +
                    "  url: window.location.href, " +
                    "  title: document.title " +
                    "}); })()",
                    new android.webkit.ValueCallback<String>() {
                        @Override
                        public void onReceiveValue(String value) {
                            JSObject result = new JSObject();
                            try {
                                String unescaped = unescapeJsString(value);
                                JSObject parsed = new JSObject(unescaped);
                                result.put("html", parsed.optString("html", ""));
                                result.put("text", parsed.optString("text", ""));
                                result.put("url", parsed.optString("url", ""));
                                result.put("title", parsed.optString("title", ""));
                            } catch (Exception e) {
                                // Fallback: return what we can from the WebView directly
                                result.put("html", "");
                                result.put("text", "");
                                result.put("url", jaalWebView.getUrl() != null
                                    ? jaalWebView.getUrl() : "");
                                result.put("title", jaalWebView.getTitle() != null
                                    ? jaalWebView.getTitle() : "");
                            }
                            call.resolve(result);
                        }
                    }
                );
            }
        });
    }

    @PluginMethod
    public void injectScript(final PluginCall call) {
        final String js = call.getString("js");
        if (js == null || js.isEmpty()) {
            call.reject("js is required");
            return;
        }

        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView == null) {
                    call.reject("WebView not initialized. Call navigate() first.");
                    return;
                }

                jaalWebView.evaluateJavascript(js,
                    new android.webkit.ValueCallback<String>() {
                        @Override
                        public void onReceiveValue(String value) {
                            JSObject result = new JSObject();
                            result.put("result", value != null ? value : "null");
                            call.resolve(result);
                        }
                    }
                );
            }
        });
    }

    @PluginMethod
    public void setPrivacyConfig(final PluginCall call) {
        PrivacyConfigHolder newConfig = new PrivacyConfigHolder();

        newConfig.blockThirdPartyCookies = call.getBoolean(
            "blockThirdPartyCookies", true
        );
        newConfig.storageIsolation = call.getBoolean(
            "storageIsolation", false
        );
        newConfig.disableJavaScript = call.getBoolean(
            "disableJavaScript", false
        );
        newConfig.blockGeolocation = call.getBoolean(
            "blockGeolocation", true
        );

        String userAgent = call.getString("userAgent");
        newConfig.userAgent = userAgent;

        // Parse blockedDomains array
        try {
            JSONArray blocked = call.getArray("blockedDomains");
            if (blocked != null) {
                newConfig.blockedDomains = new ArrayList<>();
                for (int i = 0; i < blocked.length(); i++) {
                    newConfig.blockedDomains.add(blocked.getString(i));
                }
            }
        } catch (JSONException ignored) {
            // Keep default empty list
        }

        // Parse readOnlyDomains array
        try {
            JSONArray readOnly = call.getArray("readOnlyDomains");
            if (readOnly != null) {
                newConfig.readOnlyDomains = new ArrayList<>();
                for (int i = 0; i < readOnly.length(); i++) {
                    newConfig.readOnlyDomains.add(readOnly.getString(i));
                }
            }
        } catch (JSONException ignored) {
            // Keep default empty list
        }

        this.privacyConfig = newConfig;

        // Apply to existing WebView if already created
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView != null) {
                    applyPrivacyConfig();
                }
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void show(final PluginCall call) {
        final boolean animate = call.getBoolean("animate", true);
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView != null) {
                    showOverlay(animate);
                }
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void hide(final PluginCall call) {
        final boolean animate = call.getBoolean("animate", true);
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView != null) {
                    hideOverlay(animate);
                }
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void destroy(final PluginCall call) {
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                destroyWebView();
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void getState(final PluginCall call) {
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                JSObject state = new JSObject();
                state.put("visible", isVisible);
                state.put("url", jaalWebView != null && jaalWebView.getUrl() != null
                    ? jaalWebView.getUrl() : "");
                state.put("loading", jaalClient != null && jaalClient.isLoading());
                call.resolve(state);
            }
        });
    }

    @PluginMethod
    public void goBack(final PluginCall call) {
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView != null && jaalWebView.canGoBack()) {
                    jaalWebView.goBack();
                }
                JSObject result = new JSObject();
                result.put("canGoBack",
                    jaalWebView != null && jaalWebView.canGoBack());
                call.resolve(result);
            }
        });
    }

    @PluginMethod
    public void goForward(final PluginCall call) {
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView != null && jaalWebView.canGoForward()) {
                    jaalWebView.goForward();
                }
                JSObject result = new JSObject();
                result.put("canGoForward",
                    jaalWebView != null && jaalWebView.canGoForward());
                call.resolve(result);
            }
        });
    }

    @PluginMethod
    public void reload(final PluginCall call) {
        getBridge().executeOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (jaalWebView != null) {
                    jaalWebView.reload();
                }
                call.resolve();
            }
        });
    }

    // ---- Event emission (called by JaalWebViewClient) ----

    /**
     * Emit 'navigationChange' event to JavaScript listeners.
     */
    void emitNavigationChange(String url, boolean isMainFrame, boolean isUserInitiated) {
        JSObject data = new JSObject();
        data.put("url", url);
        data.put("isMainFrame", isMainFrame);
        data.put("isUserInitiated", isUserInitiated);
        notifyListeners("navigationChange", data);
    }

    /**
     * Emit 'pageLoad' event to JavaScript listeners.
     */
    void emitPageLoad(String url, String title) {
        JSObject data = new JSObject();
        data.put("url", url);
        data.put("title", title != null ? title : "");
        notifyListeners("pageLoad", data);
    }

    /**
     * Emit 'pageError' event to JavaScript listeners.
     */
    void emitPageError(String url, String error, int errorCode) {
        JSObject data = new JSObject();
        data.put("url", url);
        data.put("error", error);
        data.put("errorCode", errorCode);
        notifyListeners("pageError", data);
    }

    /**
     * Emit 'requestBlocked' event to JavaScript listeners.
     */
    void emitRequestBlocked(String url, String matchedDomain) {
        JSObject data = new JSObject();
        data.put("url", url);
        data.put("matchedDomain", matchedDomain);
        notifyListeners("requestBlocked", data);
    }

    // ---- Internal helpers ----

    private void showOverlay(boolean animate) {
        if (jaalWebView == null) return;
        jaalWebView.animate().cancel();
        if (animate) {
            jaalWebView.setAlpha(0f);
            jaalWebView.setVisibility(View.VISIBLE);
            jaalWebView.animate().alpha(1f).setDuration(200).start();
        } else {
            jaalWebView.setVisibility(View.VISIBLE);
        }
        isVisible = true;
    }

    private void hideOverlay(boolean animate) {
        if (jaalWebView == null) return;
        jaalWebView.animate().cancel();
        if (animate) {
            jaalWebView.animate().alpha(0f).setDuration(200)
                .withEndAction(new Runnable() {
                    @Override
                    public void run() {
                        if (jaalWebView != null) {
                            jaalWebView.setVisibility(View.GONE);
                        }
                        isVisible = false;
                    }
                }).start();
        } else {
            jaalWebView.setVisibility(View.GONE);
            isVisible = false;
        }
    }

    /**
     * Fully destroy the WebView and release all resources.
     * Safe to call when already null.
     */
    private void destroyWebView() {
        if (jaalWebView != null) {
            jaalWebView.stopLoading();
            jaalWebView.setWebViewClient(null);
            jaalWebView.setWebChromeClient(null);
            ViewGroup parent = (ViewGroup) jaalWebView.getParent();
            if (parent != null) {
                parent.removeView(jaalWebView);
            }
            jaalWebView.destroy();
            jaalWebView = null;
            jaalClient = null;
            isVisible = false;
        }
    }

    /**
     * Apply the current privacy config to the existing WebView.
     * Called after setPrivacyConfig() when the WebView is already created.
     */
    private void applyPrivacyConfig() {
        if (jaalWebView == null) return;

        WebSettings settings = jaalWebView.getSettings();
        settings.setJavaScriptEnabled(!privacyConfig.disableJavaScript);
        settings.setGeolocationEnabled(!privacyConfig.blockGeolocation);

        CookieManager.getInstance().setAcceptThirdPartyCookies(
            jaalWebView, !privacyConfig.blockThirdPartyCookies
        );

        if (privacyConfig.userAgent != null && !privacyConfig.userAgent.isEmpty()) {
            settings.setUserAgentString(privacyConfig.userAgent);
        }

        // Update the client's config reference for request interception
        if (jaalClient != null) {
            jaalClient.updatePrivacyConfig(privacyConfig);
        }
    }

    /**
     * Unescape a JSON string returned by WebView.evaluateJavascript().
     *
     * Android's evaluateJavascript wraps the result in extra double quotes
     * and escapes internal quotes. This method undoes that wrapping.
     *
     * Example input: "\"{ \\\"html\\\": \\\"<html>...\\\" }\""
     * Example output: "{ \"html\": \"<html>...\" }"
     */
    static String unescapeJsString(String value) {
        if (value == null || "null".equals(value)) return "{}";
        // Remove surrounding quotes if present
        if (value.startsWith("\"") && value.endsWith("\"")) {
            value = value.substring(1, value.length() - 1);
        }
        return value
            .replace("\\\"", "\"")
            .replace("\\\\", "\\")
            .replace("\\n", "\n")
            .replace("\\t", "\t");
    }
}

File: BEYLE MOBILE/plugins/jaal-browser/ios/Plugin/JaalBrowserPlugin.swift
// BEYLE MOBILE/plugins/jaal-browser/ios/Plugin/JaalBrowserPlugin.swift
// Slice S20: iOS Capacitor plugin for the JAAL Research Browser.
//
// Manages a native WKWebView overlaid on top of the Capacitor WebView.
// Provides full browser control for JAAL research browsing: navigation,
// content extraction, JavaScript injection, and privacy enforcement
// (content blocking via WKContentRuleList, storage isolation via
// WKWebsiteDataStore, geolocation blocking).
//
// The WKWebView is lazily created on the first navigate() call and added
// as a subview of the view controller's view above Capacitor's WebView.
// It starts hidden and is shown/hidden via show()/hide() or automatically
// on navigate().
//
// All @objc methods that touch the WebView are dispatched to the main
// thread — WKWebView MUST be manipulated on the main thread only.
//
// iOS-specific constraints (documented in work scope, not bugs):
//   - Request interception limited to custom URL schemes (Apple mandate)
//   - Cookie control is per-store, not per-cookie
//   - Fingerprint defense limited by WebKit API surface (~80% vs desktop)
//   - Compensated with WKContentRuleList for broad tracker/ad blocking

import Foundation
import WebKit
import Capacitor

/// JAAL Browser Capacitor Plugin — Slice S20.
///
/// Exposes a native iOS WKWebView with privacy control to the
/// JAAL React UI components via the Capacitor JS bridge.
@objc(JaalBrowserPlugin)
public class JaalBrowserPlugin: CAPPlugin, CAPBridgedPlugin {

    // MARK: - CAPBridgedPlugin conformance

    public let identifier = "JaalBrowserPlugin"
    public let jsName = "JaalBrowser"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "navigate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPageContent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "injectScript", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setPrivacyConfig", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hide", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "destroy", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "goBack", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "goForward", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reload", returnType: CAPPluginReturnPromise)
    ]

    // MARK: - State

    private var jaalWebView: WKWebView?
    private var jaalDelegate: JaalWebViewController?
    private var privacyConfig = PrivacyConfig()
    private var isVisible: Bool = false

    // MARK: - Plugin lifecycle

    override public func load() {
        privacyConfig = PrivacyConfig()
    }

    // MARK: - Lazy WKWebView creation

    /// Create and configure the native WKWebView overlay if it doesn't already exist.
    /// MUST be called on the main thread.
    ///
    /// Layout hierarchy:
    ///   viewController.view
    ///     └─ Capacitor's WKWebView (index 0)
    ///     └─ JAAL WKWebView (added on top, starts hidden)
    private func ensureWebView() {
        guard jaalWebView == nil else { return }

        // --- WKWebViewConfiguration ---
        let config = WKWebViewConfiguration()

        // Privacy: use non-persistent data store when storage isolation is enabled,
        // otherwise use default store (persists cookies/cache across sessions).
        if privacyConfig.storageIsolation {
            config.websiteDataStore = WKWebsiteDataStore.nonPersistent()
        }

        // Preferences — JavaScript enabled/disabled
        // WKPreferences.javaScriptEnabled was deprecated in iOS 14.
        // Use WKWebpagePreferences.allowsContentJavaScript on iOS 14+.
        if #available(iOS 14.0, *) {
            config.defaultWebpagePreferences.allowsContentJavaScript = !privacyConfig.disableJavaScript
        } else {
            config.preferences.javaScriptEnabled = !privacyConfig.disableJavaScript
        }

        config.preferences.javaScriptCanOpenWindowsAutomatically = false

        // --- Create WKWebView ---
        let webView = WKWebView(frame: .zero, configuration: config)
        webView.translatesAutoresizingMaskIntoConstraints = false
        webView.isHidden = true
        webView.backgroundColor = .white
        webView.scrollView.backgroundColor = .white

        // Accessibility
        webView.accessibilityLabel = "JAAL Research Browser"

        // Custom User-Agent for fingerprint defense
        if let ua = privacyConfig.userAgent, !ua.isEmpty {
            webView.customUserAgent = ua
        }

        // --- Navigation delegate ---
        jaalDelegate = JaalWebViewController(plugin: self, privacyConfig: privacyConfig)
        webView.navigationDelegate = jaalDelegate

        // --- Compile content blocking rules ---
        jaalDelegate?.compileContentRuleList(for: webView)

        // --- Add as subview above Capacitor's WebView ---
        guard let parentView = bridge?.viewController?.view else {
            return
        }
        parentView.addSubview(webView)

        // Full-screen constraints (respects safe area on notched devices)
        NSLayoutConstraint.activate([
            webView.topAnchor.constraint(equalTo: parentView.safeAreaLayoutGuide.topAnchor),
            webView.leadingAnchor.constraint(equalTo: parentView.leadingAnchor),
            webView.trailingAnchor.constraint(equalTo: parentView.trailingAnchor),
            webView.bottomAnchor.constraint(equalTo: parentView.bottomAnchor)
        ])

        jaalWebView = webView
    }

    // MARK: - @objc Plugin Methods

    @objc func navigate(_ call: CAPPluginCall) {
        guard let urlString = call.getString("url"), !urlString.isEmpty else {
            call.reject("url is required")
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.ensureWebView()

            guard let webView = self.jaalWebView,
                  let url = URL(string: urlString) else {
                call.reject("Invalid URL: \(urlString)")
                return
            }

            webView.load(URLRequest(url: url))
            self.showOverlay(animate: true)
            call.resolve()
        }
    }

    @objc func getPageContent(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let webView = self.jaalWebView else {
                call.resolve([
                    "html": "",
                    "text": "",
                    "url": "",
                    "title": ""
                ])
                return
            }

            // Extract HTML, text, URL, and title via JavaScript evaluation
            let js = """
            (function() {
                return JSON.stringify({
                    html: document.documentElement.outerHTML,
                    text: document.body ? document.body.innerText : '',
                    url: window.location.href,
                    title: document.title
                });
            })()
            """

            webView.evaluateJavaScript(js) { result, error in
                if let error = error {
                    // Fallback: return what we can from the WebView directly
                    call.resolve([
                        "html": "",
                        "text": "",
                        "url": webView.url?.absoluteString ?? "",
                        "title": webView.title ?? ""
                    ])
                    return
                }

                guard let jsonString = result as? String,
                      let data = jsonString.data(using: .utf8),
                      let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                    // Parse failed — return fallback
                    call.resolve([
                        "html": "",
                        "text": "",
                        "url": webView.url?.absoluteString ?? "",
                        "title": webView.title ?? ""
                    ])
                    return
                }

                call.resolve([
                    "html": parsed["html"] as? String ?? "",
                    "text": parsed["text"] as? String ?? "",
                    "url": parsed["url"] as? String ?? "",
                    "title": parsed["title"] as? String ?? ""
                ])
            }
        }
    }

    @objc func injectScript(_ call: CAPPluginCall) {
        guard let js = call.getString("js"), !js.isEmpty else {
            call.reject("js is required")
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let webView = self.jaalWebView else {
                call.reject("WebView not initialized. Call navigate() first.")
                return
            }

            webView.evaluateJavaScript(js) { result, error in
                if let error = error {
                    call.reject("Script execution failed: \(error.localizedDescription)")
                    return
                }

                let resultString: String
                if let result = result {
                    if let str = result as? String {
                        resultString = str
                    } else if let data = try? JSONSerialization.data(
                        withJSONObject: result
                    ) {
                        resultString = String(data: data, encoding: .utf8) ?? "null"
                    } else {
                        resultString = "\(result)"
                    }
                } else {
                    resultString = "null"
                }

                call.resolve(["result": resultString])
            }
        }
    }

    @objc func setPrivacyConfig(_ call: CAPPluginCall) {
        var newConfig = PrivacyConfig()

        newConfig.blockThirdPartyCookies = call.getBool("blockThirdPartyCookies") ?? true
        newConfig.storageIsolation = call.getBool("storageIsolation") ?? false
        newConfig.disableJavaScript = call.getBool("disableJavaScript") ?? false
        newConfig.blockGeolocation = call.getBool("blockGeolocation") ?? true
        newConfig.userAgent = call.getString("userAgent")

        // Parse blockedDomains array
        if let blocked = call.getArray("blockedDomains") as? [String] {
            newConfig.blockedDomains = blocked
        }

        // Parse readOnlyDomains array
        if let readOnly = call.getArray("readOnlyDomains") as? [String] {
            newConfig.readOnlyDomains = readOnly
        }

        self.privacyConfig = newConfig

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            if self.jaalWebView != nil {
                self.applyPrivacyConfig()
            }
            call.resolve()
        }
    }

    @objc func show(_ call: CAPPluginCall) {
        let animate = call.getBool("animate") ?? true
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            if self.jaalWebView != nil {
                self.showOverlay(animate: animate)
            }
            call.resolve()
        }
    }

    @objc func hide(_ call: CAPPluginCall) {
        let animate = call.getBool("animate") ?? true
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            if self.jaalWebView != nil {
                self.hideOverlay(animate: animate)
            }
            call.resolve()
        }
    }

    @objc func destroy(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.destroyWebView()
            call.resolve()
        }
    }

    @objc func getState(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            call.resolve([
                "visible": self.isVisible,
                "url": self.jaalWebView?.url?.absoluteString ?? "",
                "loading": self.jaalDelegate?.isLoading ?? false
            ])
        }
    }

    @objc func goBack(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            if let webView = self.jaalWebView, webView.canGoBack {
                webView.goBack()
            }
            call.resolve([
                "canGoBack": self.jaalWebView?.canGoBack ?? false
            ])
        }
    }

    @objc func goForward(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            if let webView = self.jaalWebView, webView.canGoForward {
                webView.goForward()
            }
            call.resolve([
                "canGoForward": self.jaalWebView?.canGoForward ?? false
            ])
        }
    }

    @objc func reload(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.jaalWebView?.reload()
            call.resolve()
        }
    }

    // MARK: - Event emission (called by JaalWebViewController)

    /// Emit 'navigationChange' event to JavaScript listeners.
    func emitNavigationChange(url: String, isMainFrame: Bool, isUserInitiated: Bool) {
        notifyListeners("navigationChange", data: [
            "url": url,
            "isMainFrame": isMainFrame,
            "isUserInitiated": isUserInitiated
        ])
    }

    /// Emit 'pageLoad' event to JavaScript listeners.
    func emitPageLoad(url: String, title: String) {
        notifyListeners("pageLoad", data: [
            "url": url,
            "title": title
        ])
    }

    /// Emit 'pageError' event to JavaScript listeners.
    func emitPageError(url: String, error: String, errorCode: Int) {
        notifyListeners("pageError", data: [
            "url": url,
            "error": error,
            "errorCode": errorCode
        ])
    }

    /// Emit 'requestBlocked' event to JavaScript listeners.
    func emitRequestBlocked(url: String, matchedDomain: String) {
        notifyListeners("requestBlocked", data: [
            "url": url,
            "matchedDomain": matchedDomain
        ])
    }

    // MARK: - Internal helpers

    private func showOverlay(animate: Bool) {
        guard let webView = jaalWebView else { return }

        if animate {
            webView.alpha = 0
            webView.isHidden = false
            UIView.animate(withDuration: 0.2) {
                webView.alpha = 1
            }
        } else {
            webView.isHidden = false
            webView.alpha = 1
        }
        isVisible = true
    }

    private func hideOverlay(animate: Bool) {
        guard let webView = jaalWebView else { return }

        if animate {
            UIView.animate(withDuration: 0.2, animations: {
                webView.alpha = 0
            }, completion: { [weak self] _ in
                webView.isHidden = true
                self?.isVisible = false
            })
        } else {
            webView.isHidden = true
            webView.alpha = 0
            isVisible = false
        }
    }

    /// Fully destroy the WKWebView and release all resources.
    /// Safe to call when already nil.
    private func destroyWebView() {
        guard let webView = jaalWebView else { return }

        webView.stopLoading()
        webView.navigationDelegate = nil

        // Remove content rule list before removing from view
        jaalDelegate?.removeContentRuleList(from: webView)

        webView.removeFromSuperview()
        jaalWebView = nil
        jaalDelegate = nil
        isVisible = false
    }

    /// Apply the current privacy config to the existing WebView.
    /// Called after setPrivacyConfig() when the WebView is already created.
    ///
    /// Note: WKWebView.configuration returns a copy — some properties
    /// (websiteDataStore, JS enabled) cannot be changed after creation.
    /// For those, we must destroy and recreate the WebView.
    private func applyPrivacyConfig() {
        guard let webView = jaalWebView else { return }

        // Check if we need to recreate the WebView (data store change).
        // WKWebView.configuration.websiteDataStore cannot be reassigned
        // after creation — it's set at WKWebViewConfiguration init time.
        let currentStoreIsNonPersistent = !webView.configuration.websiteDataStore.isPersistent
        let needsNonPersistent = privacyConfig.storageIsolation
        if currentStoreIsNonPersistent != needsNonPersistent {
            // Must recreate WebView with new configuration.
            let currentUrl = webView.url
            destroyWebView()
            ensureWebView()
            // Reload the previous URL if any
            if let url = currentUrl, let newWebView = jaalWebView {
                newWebView.load(URLRequest(url: url))
                showOverlay(animate: false)
            }
            return
        }

        // Update custom User-Agent (directly settable on live WebView)
        if let ua = privacyConfig.userAgent, !ua.isEmpty {
            webView.customUserAgent = ua
        }

        // Update navigation delegate's config reference
        jaalDelegate?.privacyConfig = privacyConfig

        // Recompile content blocking rules with new blockedDomains
        jaalDelegate?.compileContentRuleList(for: webView)
    }
}
